Add tournament streaming and management toolchain

peppy opened the pull request:
PRing to potentially get some feedback on this. Is a big change but only going to grow over the week as I am for functional completion to replace current tournament streaming UI in preparation for OWC.

Remaining tasks:

- [x] Add date entry
- Fix zoom logic
- Move as much event bind/reactive logic to non-drawable classes as possible
- [x] Add proper non-visualtests project and remove menu shortcut for drawings
- [x] Depends on ppy/osu-framework#2547.
- [x] Depends on #5052.
- [x] Depends on #5054.

~Sample bracket: https://gist.github.com/peppy/0523577997de3c698388c11152e62f60~ (outdated format)

peppy committed:
Use local framework

peppy committed:
Add ability to change best of, add and delete groupings

peppy committed:
Fix incorrect access definitions

peppy committed:
Fix not being able to decrement scores of matches with no defined progression

peppy committed:
Fix transfer of teams in the case loser and winner progression are equal

peppy committed:
fixup! Fix incorrect access definitions

peppy committed:
Add team intro screen

Also adds dates to groups and matches (must be manually populated via json)

peppy committed:
Allow dynamic recompilation of beatmap panel testcase

peppy committed:
Add map pool beatmaps to groupings

peppy committed:
Add WIP map pool testcase

peppy committed:
Populate beatmaps with api information when not present

peppy committed:
Make map pool layout more correct

peppy committed:
Fix filename mismatch

peppy committed:
Fix unused properties

peppy committed:
Don't cache manager

peppy committed:
Add basic scene manager

peppy committed:
Merge branch 'master' into tournament-tools

peppy committed:
Change grouping title colours to match white background

peppy committed:
Add test videos and adjust alignment to match

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Reduce test case crashes when missing data is present

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Fix API regression in direct

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Update drawings design

peppy committed:
Add drawings screen

peppy committed:
Read and write bracket from storage instead of arbitrary location

peppy committed:
Fix filename

peppy committed:
Add showcase screen

peppy committed:
Read flag name from file

peppy committed:
Read videos from storage

peppy committed:
Read from stable

peppy committed:
Use logo-less video by default

peppy committed:
Merge tag '2018.1105.0' into tournament-tools

peppy committed:
Allow running tournament tools from main executable

peppy committed:
Remove now unnecessary null setters

peppy committed:
Hide game cursor so it is not included in stream

peppy committed:
Add missing header

peppy committed:
Add rider config for tournament run

peppy committed:
Allow tests to run custom game

peppy committed:
Fix regression in testing logic

peppy committed:
Add skeleton for gameplay screen

peppy committed:
Automatically keep window wide enough to display correctly

peppy committed:
Add the concept of "current match"

peppy committed:
Rearrange scenes

peppy committed:
Perform mappings earlier in execution

peppy committed:
Add custom font

peppy committed:
Add ability to select current match

peppy committed:
Add automatic country/user information lookups

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Move all IPC handling to its own class

peppy committed:
Move control panel logic to its own class

peppy committed:
Add support for picks and bans

peppy committed:
Automate picks and bans from IPC

peppy committed:
Improve appearance

peppy committed:
Safety check

peppy committed:
Fix non-unbinding bindable bind

peppy committed:
Fix old maps not getting cleared when switching matches

peppy committed:
Expose LargeTextureStore

peppy committed:
Add team and star displays

peppy committed:
Add non-working warmup toggle button

peppy committed:
Fill out username from user ids, rather than user ids from user ids

peppy committed:
Fix debug tools not working

peppy committed:
Fix context menus (temporarily)

peppy committed:
Fix ladder manager not saving changes out correctly

peppy committed:
Populate json with enums better

peppy committed:
Add ability to adjust scores from gameplay screen

peppy committed:
Fix many unnecessary requests being fired

peppy committed:
Fix chat making needless requests

peppy committed:
Share header area between screens

peppy committed:
Add automatic scoring

peppy committed:
Fix incorrect order after bans

peppy committed:
Animate song bar based on game mode

peppy committed:
Avoid rendering more than one video at once

peppy committed:
Handle delayed contract

peppy committed:
Fix date display

peppy committed:
Fix missing logo

peppy committed:
Fix warmup state, automate more

peppy committed:
Adjust gameplay position ever so slightly

peppy committed:
Add schedule screen

peppy committed:
Add win screen

peppy committed:
Flash panels on select

peppy committed:
Add CS display

peppy committed:
Fix regression in song bar display logic

peppy committed:
Merge branch 'master' into tournament-tools

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Fix song length not being rounded correctly when DT is applied

peppy committed:
Fix star colours being incorrect

peppy committed:
Update to support new dropdown logic

peppy committed:
Don't crash on unexpected pairing links

peppy committed:
Add score display

peppy committed:
Add back padding

peppy committed:
Add tournament chat display

peppy committed:
Fix masking of song bar

peppy committed:
Add chat IPC and gameplay screen integration

peppy committed:
Adjust spacing on map pool screen

peppy committed:
Move editing functionality to its own screen

peppy committed:
Expose groupings editor

peppy committed:
Fix code sanity issues

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Make chat tests work again, clear old messages

peppy committed:
Add date entry for groupings

peppy committed:
Add date entry for pairings

peppy committed:
Don’t hard fail if stable install is missing

peppy committed:
Show controls above screens

peppy committed:
Add better date string formatting

peppy committed:
Don't show matches on schedule which don't have teams set

peppy committed:
Show times on schedule

peppy committed:
Add logo to ladder screen

peppy committed:
Use already populated beatmap values if available

peppy committed:
Fix chat not appearing at ranking

peppy committed:
Colour chat names

peppy committed:
Actually add match score display to gameplay screen

peppy committed:
Add mod icons

peppy committed:
Group map pool by mod type

peppy committed:
Move channel reading to new ipc file

peppy committed:
Centralise chat display

peppy committed:
Use local framework

peppy committed:
Limit horizontal bounds of score display

peppy committed:
Reduce rate of scale of score display

peppy committed:
Add automation of map pool and win screen workflows

peppy committed:
Automate return to gameplay after map pool selection

peppy committed:
Add missing space after "CS"

peppy committed:
Change precedence of chat colouring

peppy committed:
Adjust gameplay screen size

peppy committed:
Don't show matches on showcase which are more than four days away

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Fix stats not displaying on showcase screen

peppy committed:
Add OD display and fix "*" difficuty modifier display

peppy committed:
Adjust gameplay area to match better

Also reduces size of score display to avoid hiding combos.

peppy committed:
Fix alignment and fonts of tem intro screen

peppy committed:
Fix spacing and colouring of song bar eleemnts

Also uses MM:ss instead of ssss.

peppy committed:
Ensure last state is set correctly

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Fix handling of API responses

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Add conditional match support

peppy committed:
Merge branch 'tournament-tools' of https://github.com/peppy/osu into tournament-tools

peppy committed:
Update in line with framework changes

peppy committed:
Automatically fix invalid pairing dates on load (based on contained groupings)

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Split out chat component into reusable piece

peppy committed:
Merge branch 'master' into tournament-tools

# Conflicts:
#	osu.Game/Online/API/Requests/Responses/APIBeatmap.cs
#	osu.Game/OsuGameBase.cs
#	osu.Game/Overlays/ChatOverlay.cs
#	osu.Game/Screens/Menu/MainMenu.cs
#	osu.Game/Screens/Tournament/Drawings.cs
#	osu.Game/Screens/Tournament/Teams/DrawingsTeam.cs
#	osu.Game/osu.Game.csproj
#	osu.sln

peppy committed:
Changes for mania world cup

peppy committed:
Add ability to reset bracket

peppy committed:
Use nuget packages again

peppy committed:
Update paths

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

# Conflicts:
#	osu.Game.Tournament.Tests/TestCaseDrawings.cs
#	osu.Game/osu.Game.csproj

peppy committed:
Update in line with screen changes

peppy committed:
Fix unnecessary texture atlas generation

peppy committed:
Fix regressed test

peppy committed:
Bind to correct target (hacky)

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Fix bindable changes

peppy committed:
Fix test bindables

peppy committed:
Fix formatting issues

peppy committed:
Fix win screen being incorrectly displayed after switching matches

peppy committed:
Fix obsolete font usages

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Enforce max consecutive blank lines

peppy committed:
Apply codefactor fixes

peppy committed:
Simplify and extract complex method

peppy committed:
Update licence headers

peppy committed:
Fix remaining inspections

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Merge branch 'move-screen-dependencies' into tournament-tools

peppy committed:
Update obsoleted font usages

smoogipoo commented on osu.Game/Overlays/Settings/SettingsItem.cs:
```diff
@@ -63,6 +63,7 @@ public virtual Bindable<T> Bindable
 
             set
             {
+                controlWithCurrent?.Current.UnbindBindings();
```
```suggestion
                if (bindable != null)
                    controlWithCurrent?.Current.UnbindFrom(bindable);
```

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Update test case naming

peppy committed:
Remove redundant code

peppy committed:
Update header colour to match TWC

peppy committed:
Hotfix to fix chat scrolling to end

peppy committed:
Make use of ValueChanged.OldValue

peppy committed:
Use less DI where we already have access to LadderInfo

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Fix Aquatico font lookups

peppy committed:
Fix regression in FileBasedIPC implementation

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Re-expose OsuButton

peppy committed:
Reduce update rate of paths

peppy committed:
Limit zoom range of bracket display

peppy committed:
Remove unnecessary invalidation

peppy commented on osu.Game.Tournament/Components/DateTextBox.cs:
```diff
@@ -0,0 +1,43 @@
+            {
+                try
+                {
+                    bindable.Value = DateTimeOffset.Parse(sender.Text);
```
newText is a bool here

peppy commented on osu.Game.Tournament/Components/DateTextBox.cs:
```diff
@@ -0,0 +1,43 @@
+                }
+                catch
+                {
+                    bindable.TriggerChange();
```
This resets the text to its previous value on a fail to parse. Will add comment.

peppy commented on osu.Game.Tournament/Components/TournamentBeatmapPanel.cs:
```diff
@@ -0,0 +1,215 @@
+            if (!string.IsNullOrEmpty(mods))
+                AddInternal(new Sprite
+                {
+                    Texture = new LargeTextureStore(new TextureLoaderStore(new StorageBackedResourceStore(storage))).Get($"mods/{mods}"),
```
no idea!

peppy commented on osu.Game.Tournament/Screens/Ladder/Components/DrawableMatchTeam.cs:
```diff
@@ -0,0 +1,206 @@
+
+        private void setCurrent()
+        {
+            //todo: tournamentgamebase?
```
no, still valid. i wanted to remove a lot of the action flow from the Drawable classes. still haven't been done yet.

peppy commented on osu.Game.Tournament/Screens/Ladder/Components/MatchPairing.cs:
```diff
@@ -0,0 +1,125 @@
+
+        public readonly Bindable<DateTimeOffset> Date = new Bindable<DateTimeOffset>();
+
+        [JsonProperty]
```
was to get around this:
![image](https://user-images.githubusercontent.com/191335/59414423-73778b00-8dfc-11e9-805a-ff3b77299bdd.png)

open to suggestion (it's only read from serialisation currently; no in-game interface

peppy commented on osu.Game.Tournament/Screens/Ladder/Components/TournamentGrouping.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public readonly BindableInt BestOf = new BindableInt(9) { Default = 9, MinValue = 3, MaxValue = 23 };
+
+        [JsonProperty]
```
Same reasoning as above. Shall I add it to all the other properties to match?

peppy commented on osu.Game.Tournament/Screens/Ladder/LadderEditorScreen.cs:
```diff
@@ -0,0 +1,171 @@
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            ((Container)InternalChild).Add(new LadderEditorSettings
```
This is specifically adding to the container created by `LadderScreen`. ie. a Composite's `Content`. I'll name and expose as Content.

peppy commented on osu.Game.Tournament/Screens/Ladder/LadderEditorScreen.cs:
```diff
@@ -0,0 +1,171 @@
+            PairingsContainer.FirstOrDefault(p => p.Pairing == pairing)?.Remove();
+        }
+
+        private class JoinRequestHandler : CompositeDrawable
```
it visualises and creates a result (the click logic results in a join being fulfilled). i'm fine with renaming though.

peppy committed:
Apply review changes

peppy committed:
Reduce size of paths

peppy committed:
Fix incorrect unbind logic

peppy committed:
Revert right click handling for now

peppy committed:
Move font local

peppy committed:
Merge branch 'master' into tournament-tools

peppy committed:
Fix crash when starting with an empty bracket

peppy committed:
Fix crash when setting a match as current which doesn't have a grouping

peppy committed:
Fix tooltips not showing at correct location

peppy committed:
Fix tests not working

peppy committed:
Fix GameplayScreen test not working

peppy committed:
Fix save button not being topmost

peppy committed:
Improve usability of grouping editor screen

peppy committed:
Fix TournamentSceneManager test

peppy committed:
Fix MapPoolScreen test

peppy committed:
TestCase -> TestScene

peppy committed:
Update grouping dropdown after new groupings are added

peppy committed:
Use bindable flow to avoid scheduled updates

peppy committed:
Use bindable logic for grouping name/description updates

peppy committed:
Create wireframe for team editor

peppy committed:
Convert TournamentTeam props to use binadbles

peppy committed:
Bindable handling in DrawableTournamentTeam

peppy committed:
Simplify LadderEditorSettings implementation by using bindables directly

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Use Current instead of Bindable

Bindable is going to be removed upstream

peppy commented:
This is mostly ready to go now.

Changes made:
- Cleaned up bindable flow in many areas
- Made sure the tools can be used from scratch
- Added UI editing of all elements (except Team's players, coming soon)
- Improved render and invalidation logic of main bracket display

These are ready to be used be non-world-cup tournament for an enthusiast user. There's plenty of UX improvements to be applied, but I'm quite happy for an initial release.

peppy committed:
Add player editor

peppy committed:
Make team editor prettier

peppy committed:
Fix start failure if videos are missing

peppy commented:
Added player editor and fixed startup issues when videos are not present.

peppy committed:
Rename and expand chat tests

peppy committed:
Undo ordering change to OsuFont

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

peppy committed:
Grouping -> Round

peppy committed:
General namespace tidy-up

peppy committed:
MatchPairing -> TournamentMatch

peppy committed:
Add xmldoc and serialisable attributes

peppy committed:
Fix mismatching filenames

peppy committed:
Remove test file

peppy committed:
Move font-related code out of TournamentGame

peppy committed:
Rename editor screens removing plurals

peppy committed:
Ignore parse errors rather than dying

peppy committed:
Mark LadderTestScene abstract

peppy committed:
Remove unused ruleset bindable

peppy committed:
Group, rename and standardise tests

peppy committed:
Move rider configurations into folders

peppy committed:
Ensure enough room for control panel display in editor tests

peppy committed:
Make common base class for editor screens

peppy committed:
Rename nested TeamRow class for clarity

peppy committed:
Add map pool editing functionality to round editor

peppy committed:
Transfer more pieces of user across

peppy committed:
Use OsuButton everywhere

peppy committed:
Add more functionality and share more logic between editor screens

peppy committed:
Add missing space

peppy committed:
Revert unnecessary changes to APIBeatmap

peppy committed:
Merge branch 'settings-item-unbind-previous-bindable' into tournament-tools

peppy committed:
Merge branch 'remote-unused-property' into tournament-tools

peppy committed:
Merge branch 'make-osu-button-instantiable' into tournament-tools

peppy commented:
- Added round-beatmap editor.
- Made a very easily extensible base class for editor screens.
- Added ability to pre-populate teams with countries.
- Added ability to clear out all teams or rounds quickly.
- Extracted all `OsuGame` changes to dependent PRs (there should now be zero line added, and only `Drawings` screen extraction resulting in removals)
- Renamed classes to be more logical (`MatchPairing` -> `TournamentMatch`; `Grouping` -> `TournamentRound`).
- Added folder grouping to rider run configurations (they were getting cluttered).
- Added folders to tournament visual tests and added any missing tests.

@smoogipoo i'm seriously done on this now. Any further changes will be in follow-up PRs.

peppy committed:
Merge branch 'master' into tournament-tools

peppy committed:
Remove migration code

peppy committed:
Merge remote-tracking branch 'upstream/master' into tournament-tools

smoogipoo committed:
Add VSCode debug configs

smoogipoo committed:
Add registry-based stable storage fallback

smoogipoo committed:
Add a max height for the ladder settings dropdowns

smoogipoo committed:
Add default date to matches

smoogipoo committed:
Fix ladder permanently disappearing after being dragged offscreen

smoogipoo commented:
@peppy I've pushed a few fixed, there's nothing super bad here and can be cleaned up later. See if you agree with my changes.

peppy committed:
Merge branch 'master' into tournament-tools
